pep-0497-externals.txt

```diff
@@ -0,0 +1,359 @@
+are associated with the named dependencies.
+
+
+Types of environment and vendor
+-------------------------------
+
+Anaconda is not an operation system. Neither is yum, neither is apt. It is the
+maintainer of the packaging system which is pertinent, and the current operating
+system is an imperfect indicator of the relevant package manager.
+
+
+Directories, categories and locations of dependency files
+---------------------------------------------------------
+
+The proposed mechanism allows introspection of the standard location for
+packages. On *nix environments, these would be (for example), /usr/bin,
+/usr/local/include etc (see https://www.gnu.org/prep/standards/html_node/Directory-Variables.html
+for a discussion of standard directories). Non-standard dependency locations
+would be supported only via the underlying platform's mechanisms for exposing
+these details, such as PATH, LD_LIBRARY_PATH etc. This PEP does not include
+generalised configuration and specification of dependencies in non-standard
+or customised locations.
+
+
+
+Alternative Syntax Concepts
+==========================
+
+Concept 0: Separated external and non-external dependencies
+-----------------------------------------------------------
+
+This is an early-stage syntax. Concept 1 replaces it as the current
+"front runner" as implementation candidate and is included in the abstract.
+
+The advantage of Concept 0 is that is clearly separates the non-Python
+dependency specification into an isolated part of the configuration arguments,
+reducing the potential for confusion between external and internal dependencies.
+This may assist during implementation as no 'sniffer' will be needed to
+determine whether a named package is internal or external.
+
+::
+
+    setup(
+        name='uses_scipy_with_atlas_extension'
+        install_requires='yaml, scipy[atlas]',
+        setup_requires='pyfortran',
+        external={
+          setup=[
+            'bin!fortran:dist_name=='ubuntu',
+            'include!kernel.h',
+            'include!liblas.h',
+          ],
+          install=[
+              'lib!liblas1.so'
+          ],
+       }
+    )
+
+Concept 1: Integrate external and non-external dependencies
+-----------------------------------------------------------
+
+A good argument for integrating these is that it will allow extensions
+to requirements.txt without needing a separate file to exist for the externals.
+It is shorter and still highly readable. Potentially instead of e.g. "lib!"
+for a prefix, "extlib!" could be used to signpost more strongly. But it's
+longer.
+
+::
+
+    setup(
+        name='uses_scipy_with_atlas_extension'
+        install_requires=['yaml', 'scipy[atlas]', 'lib!libblas1.so'],
+        setup_requires=[
+            'pyfortran',
+            'bin!fortran:dist_name=="ubuntu"',
+            'include!kernel.h',
+            'include!liblas.h',
+          ],
+    )
+
```

Comment:
Another advantage to this approach it that any tool which treats setup_requires (et al) as a list of opaque strings should "just work", while a separate field causes problems.

Notably, this change should mostly be forwards compatible with the overall structure of PEP 426.

Reply:
Revisiting this idea several years later for https://github.com/pypa/pipenv/issues/2082, and I think if we ever pursue this, we're going to need to go with the "separate fields for external dependencies" approach. The problem is that there are now plenty of tools that expect requirements definitions to comply with the PEP 508 dependency specifier format, and "failed to do anything useful with the supplied external requirements definition" is a more graceful degradation to the status quo behaviour than "threw an exception complaining about a non-compliant dependency specifier".